File: apps/ios/SearchAssistant/Services/ApiClient.swift
import Foundation

enum ApiError: Error, CustomStringConvertible {
    case invalidResponse
    case status(Int, Data?)
    case decoding(Error)
    case network(Error)

    var description: String {
        switch self {
        case .invalidResponse: return "Invalid HTTP response"
        case .status(let code, _): return "HTTP \(code)"
        case .decoding(let e): return "Decode failed: \(e)"
        case .network(let e): return "Network: \(e.localizedDescription)"
        }
    }
}

struct ApiClient {
    static let shared = ApiClient(baseURL: AppConfig.apiBaseURL)

    let baseURL: URL
    let session: URLSession

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    // MARK: - Search lifecycle

    func createSearch(_ body: CreateSearchRequest) async throws -> CreateSearchResponse {
        try await request(.post, "/api/searches", body: body)
    }

    func getSearch(slug: String) async throws -> SearchSnapshotDto {
        try await request(.get, "/api/searches/\(slug)")
    }

    func joinSearch(slug: String, displayName: String) async throws -> JoinResponse {
        try await request(.post, "/api/searches/\(slug)/join",
                          body: JoinRequest(displayName: displayName))
    }

    func me(slug: String, sessionToken: String) async throws -> ParticipantDto {
        try await request(.get, "/api/searches/\(slug)/me",
                          sessionToken: sessionToken)
    }

    // MARK: - Areas / paths (participant-authed)

    func addArea(slug: String,
                 geometry: PolygonGeometry,
                 title: String?,
                 color: String?,
                 sessionToken: String) async throws -> AreaDto {
        try await request(.post, "/api/searches/\(slug)/areas",
                          body: AddAreaRequest(geometry: geometry, title: title, color: color),
                          sessionToken: sessionToken)
    }

    func removeArea(slug: String,
                    areaId: UUID,
                    sessionToken: String) async throws {
        try await requestVoid(.delete, "/api/searches/\(slug)/areas/\(areaId.uuidString.lowercased())",
                              sessionToken: sessionToken)
    }

    func startPath(slug: String,
                   points: [[Double]],
                   sessionToken: String) async throws -> PathDto {
        try await request(.post, "/api/searches/\(slug)/paths",
                          body: StartPathRequest(points: points),
                          sessionToken: sessionToken)
    }

    func appendToPath(slug: String,
                      pathId: UUID,
                      points: [[Double]],
                      sessionToken: String) async throws -> PathDto {
        try await request(.patch, "/api/searches/\(slug)/paths/\(pathId.uuidString.lowercased())",
                          body: UpdatePathRequest(points: points, finalize: nil),
                          sessionToken: sessionToken)
    }

    func finalizePath(slug: String,
                      pathId: UUID,
                      sessionToken: String) async throws -> PathDto {
        try await request(.patch, "/api/searches/\(slug)/paths/\(pathId.uuidString.lowercased())",
                          body: UpdatePathRequest(points: nil, finalize: true),
                          sessionToken: sessionToken)
    }

    // MARK: - Manage (owner-authed)

    func updateSearch(slug: String,
                      title: String?,
                      expiresAt: Date?,
                      ownerToken: String) async throws -> UpdateSearchResponse {
        try await request(.patch, "/api/searches/\(slug)",
                          body: UpdateSearchRequest(title: title, expiresAt: expiresAt),
                          ownerToken: ownerToken)
    }

    func deleteSearch(slug: String, ownerToken: String) async throws {
        try await requestVoid(.delete, "/api/searches/\(slug)", ownerToken: ownerToken)
    }

    func clearPaths(slug: String, ownerToken: String) async throws -> ClearPathsResponse {
        try await request(.delete, "/api/searches/\(slug)/paths", ownerToken: ownerToken)
    }

    // MARK: - Plumbing

    private enum HTTPMethod: String { case get = "GET", post = "POST", patch = "PATCH", delete = "DELETE" }

    private func makeRequest(_ method: HTTPMethod,
                             _ path: String,
                             body: (any Encodable)? = nil,
                             sessionToken: String? = nil,
                             ownerToken: String? = nil) throws -> URLRequest {
        var components = URLComponents(url: baseURL.appendingPathComponent(""), resolvingAgainstBaseURL: false)!
        components.path = path
        guard let url = components.url else {
            throw URLError(.badURL)
        }
        var req = URLRequest(url: url)
        req.httpMethod = method.rawValue
        req.addValue("application/json", forHTTPHeaderField: "Content-Type")
        if let t = sessionToken { req.addValue(t, forHTTPHeaderField: "X-Session-Token") }
        if let t = ownerToken { req.addValue(t, forHTTPHeaderField: "X-Owner-Token") }
        if let body = body {
            req.httpBody = try Self.encoder.encode(AnyEncodable(body))
        }
        return req
    }

    private func request<T: Decodable>(_ method: HTTPMethod,
                                       _ path: String,
                                       body: (any Encodable)? = nil,
                                       sessionToken: String? = nil,
                                       ownerToken: String? = nil) async throws -> T {
        let req = try makeRequest(method, path,
                                  body: body,
                                  sessionToken: sessionToken,
                                  ownerToken: ownerToken)
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: req)
        } catch {
            throw ApiError.network(error)
        }
        guard let http = response as? HTTPURLResponse else { throw ApiError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            throw ApiError.status(http.statusCode, data)
        }
        do {
            return try Self.decoder.decode(T.self, from: data)
        } catch {
            throw ApiError.decoding(error)
        }
    }

    private func requestVoid(_ method: HTTPMethod,
                             _ path: String,
                             body: (any Encodable)? = nil,
                             sessionToken: String? = nil,
                             ownerToken: String? = nil) async throws {
        let req = try makeRequest(method, path,
                                  body: body,
                                  sessionToken: sessionToken,
                                  ownerToken: ownerToken)
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: req)
        } catch {
            throw ApiError.network(error)
        }
        guard let http = response as? HTTPURLResponse else { throw ApiError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            throw ApiError.status(http.statusCode, data)
        }
    }

    // MARK: - JSON encoders / decoders

    static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.dateEncodingStrategy = .custom { date, encoder in
            var container = encoder.singleValueContainer()
            try container.encode(isoFormatter.string(from: date))
        }
        return e
    }()

    static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let str = try container.decode(String.self)
            if let date = isoFormatter.date(from: str) { return date }
            if let date = isoFormatterPlain.date(from: str) { return date }
            throw DecodingError.dataCorruptedError(in: container,
                debugDescription: "Unparseable date: \(str)")
        }
        return d
    }()

    // .NET writes 7-digit fractional seconds with "+00:00" offset. Both forms
    // are accepted by ISO8601DateFormatter when .withFractionalSeconds is set.
    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoFormatterPlain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()
}

// Type-erased Encodable so we can store heterogeneous bodies in a single
// `body: (any Encodable)?` parameter. JSONEncoder still walks through to
// the underlying type's encoding.
private struct AnyEncodable: Encodable {
    private let _encode: (Encoder) throws -> Void
    init<T: Encodable>(_ wrapped: T) { self._encode = wrapped.encode }
    func encode(to encoder: Encoder) throws { try _encode(encoder) }
}

File: apps/web/src/lib/apiClient.ts
import type { Polygon } from 'geojson'
import type {
  AreaDto,
  CreateSearchRequest,
  CreateSearchResponse,
  JoinRequest,
  JoinResponse,
  ParticipantDto,
  PathDto,
  SearchSnapshotDto,
} from '../types/api'

export const API_BASE = import.meta.env.VITE_API_BASE ?? 'http://localhost:5080'

export class ApiError extends Error {
  status: number
  body?: unknown
  constructor(status: number, message: string, body?: unknown) {
    super(message)
    this.status = status
    this.body = body
  }
}

async function request<T>(path: string, init: RequestInit = {}, sessionToken?: string): Promise<T> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...(init.headers as Record<string, string> ?? {}),
  }
  if (sessionToken) headers['X-Session-Token'] = sessionToken

  const res = await fetch(API_BASE + path, { ...init, headers })
  if (!res.ok) {
    let body: unknown
    try { body = await res.json() } catch { body = await res.text() }
    throw new ApiError(res.status, `${init.method ?? 'GET'} ${path} → ${res.status}`, body)
  }
  if (res.status === 204) return undefined as T
  return (await res.json()) as T
}

export const api = {
  createSearch(body: CreateSearchRequest) {
    return request<CreateSearchResponse>('/api/searches', {
      method: 'POST', body: JSON.stringify(body),
    })
  },
  getSearch(slug: string) {
    return request<SearchSnapshotDto>(`/api/searches/${encodeURIComponent(slug)}`)
  },
  joinSearch(slug: string, body: JoinRequest) {
    return request<JoinResponse>(`/api/searches/${encodeURIComponent(slug)}/join`, {
      method: 'POST', body: JSON.stringify(body),
    })
  },
  me(slug: string, sessionToken: string) {
    return request<ParticipantDto>(`/api/searches/${encodeURIComponent(slug)}/me`,
      {}, sessionToken)
  },
  addArea(
    slug: string,
    body: { geometry: Polygon; title?: string | null; color?: string | null },
    sessionToken: string,
  ) {
    return request<AreaDto>(`/api/searches/${encodeURIComponent(slug)}/areas`, {
      method: 'POST', body: JSON.stringify(body),
    }, sessionToken)
  },
  removeArea(slug: string, areaId: string, sessionToken: string) {
    return request<void>(`/api/searches/${encodeURIComponent(slug)}/areas/${areaId}`, {
      method: 'DELETE',
    }, sessionToken)
  },
  startPath(slug: string, points: [number, number][], sessionToken: string) {
    return request<PathDto>(`/api/searches/${encodeURIComponent(slug)}/paths`, {
      method: 'POST', body: JSON.stringify({ points }),
    }, sessionToken)
  },
  appendToPath(slug: string, pathId: string, points: [number, number][], sessionToken: string) {
    return request<PathDto>(`/api/searches/${encodeURIComponent(slug)}/paths/${pathId}`, {
      method: 'PATCH', body: JSON.stringify({ points }),
    }, sessionToken)
  },
  finalizePath(slug: string, pathId: string, sessionToken: string) {
    return request<PathDto>(`/api/searches/${encodeURIComponent(slug)}/paths/${pathId}`, {
      method: 'PATCH', body: JSON.stringify({ finalize: true }),
    }, sessionToken)
  },
  updateSearch(slug: string, body: { title?: string; expiresAt?: string | null }, ownerToken: string) {
    return ownerRequest<{ title: string; expiresAt: string | null }>(
      `/api/searches/${encodeURIComponent(slug)}`,
      { method: 'PATCH', body: JSON.stringify(body) },
      ownerToken,
    )
  },
  deleteSearch(slug: string, ownerToken: string) {
    return ownerRequest<void>(
      `/api/searches/${encodeURIComponent(slug)}`,
      { method: 'DELETE' },
      ownerToken,
    )
  },
  clearPaths(slug: string, ownerToken: string) {
    return ownerRequest<{ cleared: number }>(
      `/api/searches/${encodeURIComponent(slug)}/paths`,
      { method: 'DELETE' },
      ownerToken,
    )
  },
}

async function ownerRequest<T>(path: string, init: RequestInit, ownerToken: string): Promise<T> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'X-Owner-Token': ownerToken,
    ...(init.headers as Record<string, string> ?? {}),
  }
  const res = await fetch(API_BASE + path, { ...init, headers })
  if (!res.ok) {
    let body: unknown
    try { body = await res.json() } catch { body = await res.text() }
    throw new ApiError(res.status, `${init.method ?? 'GET'} ${path} → ${res.status}`, body)
  }
  if (res.status === 204) return undefined as T
  return (await res.json()) as T
}

export const HUB_URL = API_BASE + (import.meta.env.VITE_HUB_PATH ?? '/hub/search')
